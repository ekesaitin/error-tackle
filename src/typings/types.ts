export type AnyObject = Record<string | number | symbol, any>

export interface TackleOptions {
  /** 上报信息的接口地址 */
  url?: string
  /** 上报信息的请求方式 */
  method?: 'GET' | 'POST' | 'IMG'
  /** 是否监听js错误（TypeError、ReferenceError等） */
  jsError?: boolean
  /** 是否监听promise错误 */
  promiseError?: boolean
  /** 是否监听资源错误 */
  resourceError?: boolean
  /** 是否监听console.error错误 */
  consoleError?: boolean
  /** 是否监听ajax错误 （xmlHttpRequest、fetch） */
  ajaxError?: boolean
  /** 是否监听vue错误 */
  vueError?: boolean
  /** vue应用实例 */
  vueApp?: VueApp | null
  /** 是否在控制台打印错误 */
  logError?: boolean
  /** 是否阻止错误继续向上传播 */
  coverError?: boolean
  /** 上报信息需要额外携带的数据 */
  extendsData?: any
  /** 捕获到错误时会执行的方法 */
  onError?: null | ((info: ErrorInfo) => void)
}

export type VueErrorHandler = (err: any, vm: AnyObject, info: string) => void

export interface VueApp {
  config: {
    errorHandler?: VueErrorHandler
  }
}

export interface VueErrorEvent {
  err: any
  vm: AnyObject
  info: string
}

export interface ErrorInfo {
  /** 错误类型 */
  type: string
  /** 捕获的error本体 */
  error: any
  /** 报错信息 */
  message: string
  /** 错误源文件 */
  source?: string
  /** 报错误所在的方法名 */
  name?: string
  /** 错误所在行 */
  lineno?: number | string
  /** 错误所在列 */
  colno?: number | string
  /** 出现错误的时间 */
  datatime?: string
  /** 自定义上报时额外携带的参数 */
  extendsData?: any
  /** 用户设备信息 */
  userAgent?: UserAgentObj
}

export type Reporter = (info: ErrorInfo) => void

export type Method = TackleOptions['method']

export interface BrowserReg {
  [Chrome: string]: RegExp
  IE: RegExp
  Firefox: RegExp
  Opera: RegExp
  Safari: RegExp
  '360': RegExp
  QQBrowswe: RegExp
}

export interface DeviceReg {
  [iPhone: string]: RegExp
  Android: RegExp
  iPad: RegExp
  Windows: RegExp
  Mac: RegExp
}

/** 用户设备信息 */
export interface UserAgentObj {
  /** 浏览器名称 */
  browserName: string
  /** 浏览器版本 */
  browserVersion: string
  /** 操作系统名称 */
  osName: string
  /** 操作系统版本 */
  osVersion: string
  /** 设备名称 */
  deviceName: string
}

export interface StackMap {
  stack: string
}

export const enum ERROR_TYPE {
  JS_ERROR = 'jsError',
  PROMISE_ERROR = 'promiseError',
  RESOURCE_ERROR = 'resourceError',
  CONSOLE_ERROR = 'consoleError',
  AJAX_ERROR = 'ajaxError',
}

export const enum AJAX_ERROR_TYPE {
  FETCH_ERROR = 'FetchError',
  XHR_ERROR = 'XHRError',
}
