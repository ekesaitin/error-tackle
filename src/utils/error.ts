import { getDateTime } from '.'

export interface ErrorInfo {
  type: string
  message: string
  datatime: string
  extendsData?: any
}

const getErrorType = ({ error }: ErrorEvent) => {
  switch (error.constructor) {
    case EvalError:
      return 'EvalError'
    case RangeError:
      return 'RangeError'
    case ReferenceError:
      return 'ReferenceError'
    case SyntaxError:
      return 'SyntaxError'
    case TypeError:
      return 'TypeError'
    case URIError:
      return 'URIError'
    default:
      return 'Error'
  }
}

function formatErrorInfo(e: ErrorEvent, extendsData: any): ErrorInfo
function formatErrorInfo(e: ErrorEvent, extendsData: any) {
  console.log(`extendsData====`, extendsData)
  let info = {} as ErrorInfo
  let datatime = getDateTime()
  if (e instanceof ErrorEvent) {
    info = {
      type: getErrorType(e),
      message: e.message,
      datatime,
      extendsData,
    }
  }

  return info
}

export { formatErrorInfo }
