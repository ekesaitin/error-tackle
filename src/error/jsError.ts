import { ERROR_TYPE } from '..'
import { Reporter, TackleOptions } from 'src/typings/types'
import { getErrorInfo } from './parseErrorInfo'

export const tackleJsError = (options: TackleOptions, reporter: Reporter) =>
  window.addEventListener('error', (e) => {
    const { coverError } = options
    coverError && e.preventDefault()
    let errorInfo = getErrorInfo(ERROR_TYPE.JS_ERROR, e)
    reporter(errorInfo)
  })
