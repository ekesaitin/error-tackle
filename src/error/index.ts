import { Reporter } from 'src/report'
import { getDateTime, isEventListenerObject, isVueApp } from 'src/utils'
import { tackleJsError } from './jsError'
import { tacklePromiseError } from './promiseError'
import { tackleResourceError } from './resourceError'
import { tackleVueError } from './vueError'

export type VueErrorHandler = (err: any, vm: AnyObject, info: string) => void
export interface VueApp {
  config: {
    errorHandler: VueErrorHandler
  }
}
export interface ErrorOptions {
  /** 是否监听js错误（TypeError、ReferenceError等） */
  jsError?: boolean
  /** 是否监听promise错误 */
  promiseError?: boolean
  /** 是否监听ajax错误 */
  ajaxError?: boolean
  /** 是否监听资源错误 */
  resourceError?: boolean
  /** 是否监听错误 */
  consoleError?: boolean
  /** 是否监听vue错误 */
  vueError?: boolean
  /** vue应用实例 */
  vueApp?: VueApp | null
  /** 是否阻止错误继续向上传播 */
  coverError?: boolean
  /** 上报信息需要额外携带的数据 */
  extendsData: any
}

let wrapedListener = false
const wrapListener = () => {
  const originAddEventListener = EventTarget.prototype.addEventListener
  EventTarget.prototype.addEventListener = function (type, listener, options) {
    const wrappedListener = function (...args: any[]) {
      try {
        // @ts-ignore
        return listener.apply(this, args)
      } catch (err) {
        throw err
      }
    }
    return originAddEventListener.call(this, type, wrappedListener, options)
  }
}

export const createErrorTackle = (options: ErrorOptions, reporter: Reporter) => {
  if (!wrapedListener) {
    wrapedListener = true
    wrapListener()
  }
  const { jsError, promiseError, resourceError, vueApp, vueError } = options
  if (jsError) tackleJsError(options, reporter)
  if (promiseError) tacklePromiseError(options, reporter)
  if (resourceError) tackleResourceError(options, reporter)
  if (vueError && isVueApp(vueApp)) tackleVueError(options, reporter)
}
