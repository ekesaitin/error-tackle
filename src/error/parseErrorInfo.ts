import { AJAX_ERROR_TYPE, ErrorInfo, ERROR_TYPE, VueErrorEvent } from 'src/typings/types'
import { getDateTime, getSourceFromStack } from 'src/utils'
import { getUserAgent } from 'src/utils/client'

const getErrInfoWithSource = (target: ErrorInfo, stack: string): ErrorInfo => {
  const source = getSourceFromStack(stack)
  if (source) {
    return {
      ...target,
      ...source,
    }
  } else {
    return target
  }
}

const getErrorType = (e: ErrorEvent | PromiseRejectionEvent) => {
  if (e instanceof ErrorEvent) return e.error?.name
  else if (e instanceof PromiseRejectionEvent) return e.type
}

const getJsError = (e: ErrorEvent): ErrorInfo => {
  return getErrInfoWithSource(
    {
      type: getErrorType(e),
      error: e.error,
      message: e.error?.message,
    },
    e.error.stack,
  )
}

const getPromiseError = (e: PromiseRejectionEvent): ErrorInfo => ({
  type: getErrorType(e),
  error: e.reason,
  message: e.reason?.message,
})

const getResourceError = (e: ErrorEvent): ErrorInfo => {
  const target = e.target as any
  let error = ` ${target.baseURI}页面 加载${target.tagName.toLowerCase()}: ${
    target?.src ?? target?.href
  } 错误`

  return {
    type: 'ResourceError',
    error,
    message: '资源加载错误',
  }
}

const getConsoleError = (e: any): ErrorInfo => {
  const info = {
    type: 'ConsoleError',
    error: e,
    message: `${window.location.href}，页面通过 console.error 抛出的错误，`,
  }

  if (e instanceof Error) {
    return getErrInfoWithSource(info, e.stack ?? '')
  } else {
    return info
  }
}

const getAjaxError = (e: any, type: AJAX_ERROR_TYPE): ErrorInfo => {
  return getErrInfoWithSource(
    {
      type,
      error: e,
      message: 'ajax error',
    },
    e.stack,
  )
}

export const getErrorInfo = (
  errorType: ERROR_TYPE | never,
  e: ErrorEvent | PromiseRejectionEvent | VueErrorEvent,
  ...args: any[]
): ErrorInfo => {
  let info = {} as ErrorInfo
  switch (errorType) {
    case ERROR_TYPE.JS_ERROR:
      info = getJsError(e as ErrorEvent)
      break
    case ERROR_TYPE.PROMISE_ERROR:
      info = getPromiseError(e as PromiseRejectionEvent)
      break
    case ERROR_TYPE.RESOURCE_ERROR:
      info = getResourceError(e as ErrorEvent)
      break
    case ERROR_TYPE.CONSOLE_ERROR:
      info = getConsoleError(e)
      break
    case ERROR_TYPE.AJAX_ERROR:
      info = getAjaxError(e, args[0])
      break
  }

  info.datatime = getDateTime()
  info.userAgent = getUserAgent()

  return info
}
