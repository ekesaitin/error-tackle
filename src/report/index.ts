import { ErrorInfo, Method, Reporter, TackleOptions } from 'src/typings/types'
import { isUrl, obj2query } from 'src/utils'

const fetchImg = (url: string, data: ErrorInfo) => {
  let query = obj2query(data)
  const gif = document.createElement('img')
  gif.src = url + query
}

const fetchData = (url: string, method: Method, info: ErrorInfo) => {
  let query = obj2query(info)
  url = method === 'GET' ? url + query : url
  fetch(url, {
    method,
    headers: {
      'Content-Type': 'application/json',
    },
    ...(method === 'POST' && { body: JSON.stringify(info) }),
  })
}

export const reportError = (url: string, method: Method, info: ErrorInfo) => {
  if (method === 'IMG') fetchImg(url, info)
  else fetchData(url, method, info)
}

export const createReporter =
  (options: TackleOptions): Reporter =>
  (info: ErrorInfo) => {
    let { url, method, onError, logError, coverError, extendsData } = options
    let report = {
      ...info,
      ...(extendsData ? { extendsData } : false),
    }
    console.log(`report====`, report)
    logError && coverError && console.log(report.error)
    if (url && isUrl(url)) reportError(url, method, report)
    onError?.(info)
  }
