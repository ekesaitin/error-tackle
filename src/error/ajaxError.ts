import { AnyObject } from 'src/typings/types'
import { ERROR_TYPE } from 'src'
import { AJAX_ERROR_TYPE, getErrorInfo } from 'src/error/parseErrorInfo'
import { TackleOptions, Reporter } from 'src/typings/types'
import { isObject, noop } from 'src/utils'

interface StackMap {
  stack: object
}

const shallowCloneError = <T extends object>(err: T): T => {
  if (!isObject(err)) {
    return err
  }
  const res = Object.create(null)
  for (let key in err) {
    res[key] = err[key]
  }
  return res
}

const handleReportError = (
  reporter: Reporter,
  type: AJAX_ERROR_TYPE,
  error: any,
  stackMap?: StackMap,
) => {
  const err = shallowCloneError(error)
  if (stackMap) {
    err.stack = stackMap.stack
  }
  const info = getErrorInfo(ERROR_TYPE.AJAX_ERROR, err, type)
  reporter(info)
}

const rewriteXmlHttpRequest = (reporter: Reporter) => {
  if (!window.XMLHttpRequest) {
    return
  }
  const interceptiveStack = {} as StackMap

  let xhrSend = XMLHttpRequest.prototype.send
  let _handleEvent = (e: any) => {
    try {
      if (e && e.currentTarget && e.currentTarget.status !== 200) {
        handleReportError(reporter, AJAX_ERROR_TYPE.XHR_ERROR, e, interceptiveStack)
      }
    } catch (err) {
      handleReportError(reporter, AJAX_ERROR_TYPE.XHR_ERROR, err, interceptiveStack)
    }
  }
  XMLHttpRequest.prototype.send = function newSend() {
    Error.captureStackTrace(interceptiveStack, newSend)
    if (this.addEventListener) {
      this.addEventListener('error', _handleEvent)
      this.addEventListener('load', _handleEvent)
      this.addEventListener('abort', _handleEvent)
    } else {
      let tempStateChange = this.onreadystatechange ?? noop
      this.onreadystatechange = function (event) {
        // @ts-ignore
        tempStateChange.apply(this, arguments)
        if (this.readyState === 4) {
          _handleEvent(event)
        }
      }
    }
    // @ts-ignore
    return xhrSend.apply(this, arguments)
  }
}

const rewriteFetch = (reporter: Reporter) => {
  if (!window.fetch) return

  let _oldFetch = window.fetch
  function newFetch(input: RequestInfo, init?: RequestInit | undefined) {
    const interceptiveStack = {} as StackMap
    Error.captureStackTrace(interceptiveStack, newFetch)

    return (
      _oldFetch
        // @ts-ignore
        .call(this, input, init)
        .then((res) => {
          if (!res.ok) {
            // setStack(res, interceptiveStack)
            handleReportError(reporter, AJAX_ERROR_TYPE.FETCH_ERROR, res, interceptiveStack)
          }
          return res
        })
        .catch(function fn(err) {
          // setStack(err, interceptiveStack)
          handleReportError(reporter, AJAX_ERROR_TYPE.FETCH_ERROR, err, interceptiveStack)
        })
    )
  }

  window.fetch = newFetch as typeof window.fetch
}

export const tackleAjaxError = (options: TackleOptions, reporter: Reporter) => {
  rewriteXmlHttpRequest(reporter)
  rewriteFetch(reporter)
}
