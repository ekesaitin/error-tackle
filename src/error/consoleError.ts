import { getErrorInfo } from 'src/error/parseErrorInfo'
import { TackleOptions, Reporter, ERROR_TYPE } from 'src/typings/types'
import { noop } from 'src/utils'

export const tackleConsoleError = (options: TackleOptions, reporter: Reporter) => {
  const consoleError = window.console?.error || noop
  const { coverError } = options
  window.console.error = function (...args: any[]) {
    args.forEach((err) => {
      setTimeout(() => {
        if (err instanceof Error) throw err
        else {
          let info = getErrorInfo(ERROR_TYPE.CONSOLE_ERROR, err)
          reporter(info)
        }
      })
    })

    !coverError && consoleError.apply(window, args)
  }
}
