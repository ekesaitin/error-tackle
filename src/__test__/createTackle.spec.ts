import { TackleOptions } from '../typings/types'
import { createTackle, defaultOptions } from '../index'
import * as error from '../error'
import * as report from '../report'

jest.mock('../error')
jest.mock('../report')

describe('createTackle', () => {
  const createErrorTackle: any = error.createErrorTackle
  const createReporter: any = report.createReporter

  test('没有options传入时', () => {
    createTackle()
    expect(createErrorTackle).toHaveBeenCalledTimes(1)
    expect(createErrorTackle.mock.calls[0][0]).toEqual(defaultOptions)
    expect(createReporter).toHaveBeenCalledTimes(1)
    expect(createReporter.mock.calls[0][0]).toEqual(defaultOptions)
  })

  test('options为非对象时', () => {
    let opts: any = 1
    createTackle(opts)
    opts = []
    createTackle(opts)
    opts = Symbol()
    createTackle(opts)
    expect(createErrorTackle).toBeCalledTimes(3)
    createErrorTackle.mock.calls.forEach((call: any[]) => {
      expect(call[0]).toEqual(defaultOptions)
    })
    expect(createReporter).toBeCalledTimes(3)
    createReporter.mock.calls.forEach((call: any[]) => {
      expect(call[0]).toEqual(defaultOptions)
    })
  })

  test('options为一个对象时', () => {
    const wholeOptions: TackleOptions = {
      url: 'xxxxx.com/xxx',
      method: 'POST',
      logError: false,
      jsError: false,
      promiseError: false,
      resourceError: false,
      consoleError: true,
      ajaxError: false,
      vueError: true,
      vueApp: {
        config: {
          errorHandler() {},
        },
      },
      coverError: false,
      extendsData: {
        token: 'xxx-xxxx-xx',
        user: {
          name: 'username',
          auth: 'userauth',
        },
      },
      onError: () => {},
    }

    const partialOptions: TackleOptions = {
      url: 'xxxxx.com/xxx',
      method: 'POST',
      extendsData: 1010101010,
    }

    createTackle({})
    createTackle(wholeOptions)
    createTackle(partialOptions)
    expect(createErrorTackle).toBeCalledTimes(3)
    expect(createErrorTackle.mock.calls[0][0]).toEqual(defaultOptions)
    expect(createErrorTackle.mock.calls[1][0]).toEqual(wholeOptions)
    expect(createErrorTackle.mock.calls[2][0]).toMatchSnapshot()
    expect(createReporter).toBeCalledTimes(3)
    expect(createReporter.mock.calls[0][0]).toEqual(defaultOptions)
    expect(createReporter.mock.calls[1][0]).toEqual(wholeOptions)
    expect(createReporter.mock.calls[2][0]).toMatchSnapshot()
  })
})
