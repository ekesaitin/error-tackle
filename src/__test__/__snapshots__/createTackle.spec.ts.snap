// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createTackle options为一个对象时 1`] = `
Object {
  "ajaxError": true,
  "consoleError": false,
  "coverError": false,
  "extendsData": 1010101010,
  "jsError": true,
  "logError": false,
  "method": "POST",
  "onError": null,
  "promiseError": true,
  "resourceError": true,
  "url": "xxxxx.com/xxx",
  "vueApp": null,
  "vueError": false,
}
`;

exports[`createTackle options为一个对象时 2`] = `
Object {
  "ajaxError": true,
  "consoleError": false,
  "coverError": false,
  "extendsData": 1010101010,
  "jsError": true,
  "logError": false,
  "method": "POST",
  "onError": null,
  "promiseError": true,
  "resourceError": true,
  "url": "xxxxx.com/xxx",
  "vueApp": null,
  "vueError": false,
}
`;
